Ensure focus is kept inside trap when focussed element is removed

kasperg opened the pull request:
focus-trap/focus-trap-react#962 has shown that if the element that is currently focused is removed from the DOM, the  focus is moved either to the document's body element or becomes `null`

When using `focus-trap` it would be appropriate to keep focus inside the trap. This is relevant if the trap is used for a modal and the content of the modal is updated but the modal stays open.

This adds an initial Cypress test to show whether this works or not.

<details>
<summary>PR Checklist</summary>
<br/>

__Please leave this checklist in your PR.__

- Source changes maintain stated browser compatibility.
- Includes updated docs demo bundle if source/docs code was changed (run `npm run demo-bundle` in your branch and include the `/docs/demo-bundle.js` file that gets generated in your PR).
- Issue being fixed is referenced.
- Unit test coverage added/updated.
- E2E (i.e. demos) test coverage added/updated.
  - ⚠️ Non-covered demos (look for `// TEST MANUALLY` comments [here](https://github.com/focus-trap/focus-trap/blob/master/docs/js/index.js)) that can't be fully tested in Cypress have been __manually__ verified.
- Typings added/updated.
- Changes do not break SSR:
  - Careful to test `typeof document/window !== 'undefined'` before using it in code that gets executed on load.
- README updated (API changes, instructions, etc.).
- Changes to dependencies explained.
- Changeset added (run `npm run changeset` locally to add one, and follow the prompts).
  - EXCEPTION: A Changeset is not required if the change does not affect any of the source files that produce the package bundle. For example, demo changes, tooling changes, test updates, or a new dev-only dependency to run tests more efficiently should not have a Changeset since it will not affect package consumers.

</details>

stefcameron commented on docs/index.html:
```diff
@@ -1135,6 +1135,34 @@ <h2 id="inert-heading">works with inert elements</h2>
+        </p>
+      </div>
+      <p>
+        <a href="https://github.com/focus-trap/focus-trap/blob/master/docs/js/default.js">View demo source <span aria-hidden="true">&gt;&gt;</span></a>
```
`default.js` should be `dom-remove.js` here

kasperg replied:
:+1: Fixed in [c8756d7](https://github.com/focus-trap/focus-trap/pull/932/commits/c8756d74df588f37055740df7f0c536d1a2c90cc)

stefcameron commented on docs/index.html:
```diff
@@ -1135,6 +1135,34 @@ <h2 id="inert-heading">works with inert elements</h2>
+
+    <div id="demo-dom-remove">
+      <h2 id="dom-remove-heading">Pulls focus when removing DOM elements</h2>
+      <p>
+      </p>
```
I don't think the title really explains it well. How about adding this explanation here: "Keeps focus in the trap when the focused element is removed from the DOM."

kasperg replied:
Agreed. I tried to give it another go in [46bf5cb](https://github.com/focus-trap/focus-trap/pull/932/commits/46bf5cb5c7309b8abd078a9a5da596f28393e73a).

stefcameron reviewed (changes requested):
Thank you for adding this handy demo and test (and for following nearby code patterns)!

stefcameron reviewed (commented):
Thanks for the updates so far.

kasperg commented:
@stefcameron I gave it a go implementing a fix. Tests are green so at least that's a start.

If you agree that this looks right I rewrite the commits a bit.

stefcameron commented:
@kasperg 
> @stefcameron I gave it a go implementing a fix. Tests are green so at least that's a start.
> 
> If you agree that this looks right I rewrite the commits a bit.

Thank you for the additional work you've done! I will check this out sometime over the next couple of days when I find a moment.

stefcameron reviewed (commented):
Thank you for your work on this! A couple of comments...

kasperg commented on index.js:
```diff
@@ -632,6 +651,16 @@ const createFocusTrap = function (elements, userOptions) {
       passive: false,
     });
 
+    if ('MutationObserver' in window) {
```
I am not that much into server-side JS, but the check seems reasonable. I have added it in [409d9dc](https://github.com/focus-trap/focus-trap/pull/932/commits/409d9dc3b027fe884ad69fa5c4966fc5633c6965).

kasperg commented on index.js:
```diff
@@ -254,8 +258,8 @@ const createFocusTrap = function (elements, userOptions) {
 
-    if (node === undefined) {
-      // option not specified: use fallback options
+    if (node === undefined || !node.isConnected) {
```
:+1: I have updated the the to use `isFocusable()` in [7f8b500](https://github.com/focus-trap/focus-trap/pull/932/commits/7f8b5001a73f1a1f908a8e2b5b266bfb51ff8434).

stefcameron reviewed (approved):
Looks great, thank you!

stefcameron reviewed (changes requested):
Actually, I think there's a gap in the case where the container elements are updated. The MutationObserver will keep observing old containers, no longer the right ones.

Specially related to this which I mentioned on the issue:

> As for the observer, I think there are two possible approaches there:
> 
> 1. It observes each trap container for node removals therein. Any time `updateContainerElements` is called, the observer is reset and restarts observing the new containers given. The callback checks the removed node against `state.mostRecentlyFocusedNode`.
> 2. It observes only the `state.mostRecentlyFocusedNode` and would be updated in `tryFocus()` and `checkFocusIn()` where that property is updated (or [the property](https://github.com/focus-trap/focus-trap/blob/master/index.js#L152) is converted into a `get()/set()` pair and the observer is updated in the setter so it doesn't matter where it gets updated).
> 
> I'm not sure how the observer reacts when the very node it's watching gets removed. Approach (1) might be the only way.

kasperg commented:
@stefcameron You are right. I totally missed the part about `updateContainerElements()`.

I updated the implementation with a simple change to fix address this in [1995abb](https://github.com/focus-trap/focus-trap/pull/932/commits/1995abb5993d647ade23430f5811401311304f3a).

However the code duplication lead me to reconsider the approach so I refactored most of it in [a832945](https://github.com/focus-trap/focus-trap/pull/932/commits/a832945456f439cc6bcd472256417e9953d92c65). Please check the commit comment for additional commentary.

If you agree with this new approach we can take it from there. If not we can continue from `1995abb`.

The PR had merge conflicts and needed a bit of cleanup so I have rebased the PR on `master`.

stefcameron commented on docs/js/index.js:
```diff
@@ -59,3 +59,5 @@ require('./with-open-web-component')();
 // http://localhost:9966/#demo-inert
 require('./inert')();
+
+require('./dom-remove')();
```
Just a tiny request to move this up to line 26 now that you've rebased. I reorganized the requires to put all the ones that are Cypress-testable before all those that aren't, just to make it easier for me to follow the "list" when doing manual testing.

stefcameron reviewed (commented):
@kasperg I like your new approach to solving the problem. Seems very nice. Your thought process/explanation for changing directions a bit was very helpful. I just have one final small request, which I think I'll just push up to your branch myself since it's just a line in the `/docs/index.js` now that you've rebased.

stefcameron reviewed (approved):
🚀

stefcameron commented:
If you're good with that, I'll merge it. 😄

kasperg commented:
👍 Thanks. Appreciate the help.

stefcameron commented:
@all-contributors add @kasperg for bug, code

stefcameron commented:
Published in focus-trap v7.4.3
